pkg/nsx/services/vpc/vpc.go

```diff
@@ -665,14 +650,97 @@ func (s *VPCService) CreateOrUpdateVPC(obj *v1alpha1.NetworkInfo) (*model.Vpc, *
 				// delete the nsx vpc object and re-create it in the next loop
 				if err := s.DeleteVPC(*newVpc.Path); err != nil {
 					log.Error(err, "cleanup VPC failed", "VPC", *createdVpc.Id)
-					return nil, nil, err
+					return nil, err
 				}
 			}
-			return nil, nil, err
+			return nil, err
+		}
+	}
+
+	return &newVpc, nil
+}
+
+func (s *VPCService) GetGatewayConnectionTypeFromConnectionPath(connectionPath string) (string, error) {
+	/* examples of connection_path:
+	   /infra/distributed-gateway-connections/gateway-101
+	   /infra/gateway-connections/tenant-1
+	*/
+	parts := strings.Split(connectionPath, "/")
+	if len(parts) != 4 || parts[1] != "infra" {
+		return "", fmt.Errorf("unexpected connectionPath %s", connectionPath)
+	}
+	return parts[2], nil
+}
+
+func (s *VPCService) ValidateGatewayConnectionStatus(nc *common.VPCNetworkConfigInfo) (bool, string, bool, error) {
+	autoSnatEnabled := false
+	var cursor *string
+	pageSize := int64(1000)
+	markedForDelete := false
+	res, err := s.NSXClient.VPCConnectivityProfilesClient.List(nc.Org, nc.NSXProject, cursor, &markedForDelete, nil, &pageSize, nil, nil)
+	err = nsxutil.NSXApiError(err)
+	if err != nil {
+		return false, "", autoSnatEnabled, err
+	}
+	var profiles []model.VpcConnectivityProfile
+	profiles = append(profiles, res.Results...)
+	for _, profile := range profiles {
+		if profile.ServiceGateway != nil && profile.ServiceGateway.NatConfig != nil && *profile.ServiceGateway.NatConfig.EnableDefaultSnat {
+			autoSnatEnabled = true
 		}
 	}
+	// Case 1: if service_gateway is enabled, Edge must have been deployed.
+	for _, profile := range profiles {
+		if profile.ServiceGateway != nil && *profile.ServiceGateway.Enable {
+			return true, "", autoSnatEnabled, nil
```

Comment:
autoSnatEnabled is the final ideal config, but could gateway connection be wrong or empty after the autoSnatEnabled is enabled?

Reply:
I saw "Edge is not deployed, service_gateway cannot be enabled, VPC can be created. After Edge is deployed, service_gateway can be enabled for the created VPC" in the doc. So you mean we cannot get "gateway connection is ready" from "Edge is deployed"?

Reply:
BTW, here what I checked is service_gateway, do you mean service_gateway=enable equals autoSnatEnabled=enabled?

Reply:
I had some verifications and confirmed that exactly when the service_gateway is enabled, the NSX still allows the connection gateway empty. I'll polish this logic, thanks for pointing it out.